### ISSUE TITLE
Probing syscall__renameat in RHEL 8 - must use PT_REGS_PARM5

### ISSUE THREAD

**Issue by craig65535** [iovisor/bcc#2597]
I created kprobes for the `renameat` and `renameat2` syscalls:

```
int syscall__renameat(struct pt_regs *ctx, int oldfd, const char __user *oldname,                                                                        
                      int newfd, const char __user *newname)                                                                                             
{                                                                                                                                                        
.
.
}

int syscall__renameat2(struct pt_regs *ctx, int oldfd, const char __user *oldname,
                       int newfd, const char __user *newname, unsigned int flags)
{  
.
.
}
```

I can read the data at `oldname` fine, but `newname` seems to be garbage. However, I can access the contents of `newname` fine by using `PT_REGS_PARM5(ctx)`.

Is there is a reason this doesn't work more intuitively? Is is because these syscalls have 4+ parameters?

Thank you

**Comment by yonghong-song** [comment]
The syscall__ interface supports up to 6 parameters on x64 platform.
Could you try to use `PT_REGS_PARAM4(ctx)` for `newname` which is supposed to be correct answer? Could you share your reproducible test case so people can take a further look at the issue if needed?

**Comment by craig65535** [comment]
Below is some C++ that reproduces the problem. I had it echo the expected/actual parameters, and the contents of PT_REGS_PARM1 through PARM5. Here is the output:
```
actual oldfd=11, expected oldfd=11
actual oldname="AnOldName", expected="AnOldName"
actual newfd=22, expected newfd=22
actual newname="H=", expected newname="ANewName" (FAIL)
parm1="" (-110041344245928)
parm2="" (0)
parm3="" (868816)
parm4="" (1574146436)
parm5="ANewName" (140722852987392)
```

I'm using kernel 4.18.0-80.11.2.el8_0.x86_64 on RHEL 8.0.

```C++
#include <iostream>

#include "bcc/BPF.h"

using namespace std;

static const string _probe_renameat = R"(
#include <uapi/linux/ptrace.h>

typedef struct event {
    int64_t oldfd;
    char oldname[32];
    int64_t newfd;
    char newname[32];
    int64_t parm1_int;
    char parm1_str[32];
    int64_t parm2_int;
    char parm2_str[32];
    int64_t parm3_int;
    char parm3_str[32];
    int64_t parm4_int;
    char parm4_str[32];
    int64_t parm5_int;
    char parm5_str[32];
} event_t;

BPF_PERF_OUTPUT(events);

int syscall__renameat(struct pt_regs *ctx, int oldfd, const char __user *oldname,
                      int newfd, const char __user *newname)
{
    event_t event = {
        .oldfd = oldfd,
        .newfd = newfd,
        .parm1_int = PT_REGS_PARM1(ctx),
        .parm2_int = PT_REGS_PARM2(ctx),
        .parm3_int = PT_REGS_PARM3(ctx),
        .parm4_int = PT_REGS_PARM4(ctx),
        .parm5_int = PT_REGS_PARM5(ctx),
    };
    bpf_probe_read_str(event.oldname, sizeof(event.oldname), oldname);
    bpf_probe_read_str(event.newname, sizeof(event.newname), newname);
    bpf_probe_read_str(event.parm1_str, sizeof(event.parm1_str), (const char __user *)PT_REGS_PARM1(ctx));
    bpf_probe_read_str(event.parm2_str, sizeof(event.parm2_str), (const char __user *)PT_REGS_PARM2(ctx));
    bpf_probe_read_str(event.parm3_str, sizeof(event.parm3_str), (const char __user *)PT_REGS_PARM3(ctx));
    bpf_probe_read_str(event.parm4_str, sizeof(event.parm4_str), (const char __user *)PT_REGS_PARM4(ctx));
    bpf_probe_read_str(event.parm5_str, sizeof(event.parm5_str), (const char __user *)PT_REGS_PARM5(ctx));
    events.perf_submit(ctx, &event, sizeof(event));
    return 0;
}
)";

/* This definition must mirror the one in the BPF program above. */
typedef struct event {
    int64_t oldfd;
    char oldname[32];
    int64_t newfd;
    char newname[32];
    int64_t parm1_int;
    char parm1_str[32];
    int64_t parm2_int;
    char parm2_str[32];
    int64_t parm3_int;
    char parm3_str[32];
    int64_t parm4_int;
    char parm4_str[32];
    int64_t parm5_int;
    char parm5_str[32];
} event_t;

typedef struct test_case {
    event_t expected_event;
    bool done;
    bool success;
} test_case_t;

static void _output(void *cb_ctx, void *data, int data_size)
{
    do {
        test_case_t *test_case = (test_case_t *)cb_ctx;
        test_case->done = true;
        if (data_size < 0 || (size_t)data_size < sizeof(event_t)) {
            cerr << "data size mismatch (" << data_size << " < expected " << sizeof(event_t) << ")" << endl;
            break;
        }
        auto event = static_cast<event_t *>(data);
        bool failed = false;
        cout << "actual oldfd=" << event->oldfd << ", expected oldfd=" << test_case->expected_event.oldfd;
        if (event->oldfd != test_case->expected_event.oldfd) {
            cout << " (FAIL)";
            failed = true;
        }
        cout << endl;
        cout << "actual oldname=\"" << event->oldname << "\", expected=\"" << test_case->expected_event.oldname << "\"";
        if (strcmp(event->oldname, test_case->expected_event.oldname)) {
            cout << " (FAIL)";
            failed = true;
        }
        cout << endl;
        cout << "actual newfd=" << event->newfd << ", expected newfd=" << test_case->expected_event.newfd;
        if (event->newfd != test_case->expected_event.newfd) {
            cout << " (FAIL)";
            failed = true;
        }
        cout << endl;
        cout << "actual newname=\"" << event->newname << "\", expected newname=\"" << test_case->expected_event.newname << "\"";
        if (strcmp(event->newname, test_case->expected_event.newname)) {
            cout << " (FAIL)";
            failed = true;
        }
        cout << endl;
        cout << "parm1=\"" << event->parm1_str << "\" (" << event->parm1_int << ")" << endl;
        cout << "parm2=\"" << event->parm2_str << "\" (" << event->parm2_int << ")" << endl;
        cout << "parm3=\"" << event->parm3_str << "\" (" << event->parm3_int << ")" << endl;
        cout << "parm4=\"" << event->parm4_str << "\" (" << event->parm4_int << ")" << endl;
        cout << "parm5=\"" << event->parm5_str << "\" (" << event->parm5_int << ")" << endl;
        test_case->success = !failed;
    } while (0);
}

int main()
{
    int ret = 1;

    do {
        ebpf::BPF bpf;
        auto init_res = bpf.init(_probe_renameat);
        if (init_res.code() != 0) {
            cerr << init_res.msg() << endl;
            break;
        }

        string fnname = bpf.get_syscall_fnname("renameat");
        auto attach_res = bpf.attach_kprobe(fnname, "syscall__renameat");
        if (attach_res.code() != 0) {
            cerr << attach_res.msg() << endl;
            break;
        }

        /* Set callback */
        test_case_t test_case = {
            .expected_event = {
                11,
                "AnOldName",
                22,
                "ANewName",
                0, "", 0, "", 0, "", 0, "", 0, "",
            },
            .done = false,
            .success = false,
        };
        auto open_res = bpf.open_perf_buffer("events", &_output, nullptr, &test_case);
        if (open_res.code() != 0) {
            cerr << open_res.msg() << endl;
            break;
        }

        cout << "Started" << endl;

        /* Generate event */
        (void)renameat(test_case.expected_event.oldfd, test_case.expected_event.oldname,
                       test_case.expected_event.newfd, test_case.expected_event.newname);

        int rc;
        do {
            cout << "Polling" << endl;
            rc = bpf.poll_perf_buffer("events");
        } while (!test_case.done && rc == 0);
        if (rc < 0) {
            cerr << "Failed to poll for events" << endl;
            break;
        }

        if (!test_case.success) {
            break;
        }
        cout << "Success" << endl;
        ret = 0;
    } while (0);

    return ret;
}
```

**Comment by craig65535** [comment]
I'm using bcc release v0.11.0 (0fa419a64e71984d42f107c210d3d3f0cc82d59a).

**Comment by craig65535** [comment]
@yonghong-song is this failing in the same way for you?

**Comment by craig65535** [comment]
I found that I don't need PT_REGS_PARM5 when using RHEL 7.6 (3.10.0-957.5.1.el7.x86_64). So this could be an issue specific to RHEL 8.

**Comment by craig65535** [comment]
It appears this was fixed in #3153
